### PR TITLE
Service with ConnectInfo

### PR DESCRIPTION
**Pull request by 82marbag**
## Motivation and Context

Business logic can't know where a request is coming from. Any request is opaque to the service handling it.

## Description

Hyper can make connection information available to the service; axum provides this feature.
This change will let customers implement:

```rust
pub async fn handler(
    input: input::Input,
    conn_info: Extension<aws_smithy_http_server::routing::ConnectInfo<std::net::SocketAddr>>,
) -> output::Output {
    // use `conn_info`
    output::Output {}
}
```

## Testing

Tested implementing the handler:
```rust
pub async fn gym_battle(
    input: input::BattleInput,
    conn_info: Extension<aws_smithy_http_server::routing::ConnectInfo<std::net::SocketAddr>>,
) -> output::BattleOutput {
    if c.0 .0.ip() == "127.0.0.1".parse::<std::net::IpAddr>().unwrap() {
        println!("Battles from the same gym are not supported");
        ...
    }
    output::BattleOutput { ... }
}
```

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rust-runtime/aws-smithy-http-server/src/routing/into_make_service_with_connect_info.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+/// Trait that connected IO resources implement and use to produce information
+/// about the connection.
+///
+/// The goal for this trait is to allow users to implement custom IO types that
+/// can still provide the same connection metadata.
+///
+/// See [`Router::into_make_service_with_connect_info`] for more details.
+///
+/// [`Router::into_make_service_with_connect_info`]: crate::routing::Router::into_make_service_with_connect_info
+pub trait Connected<T>: Clone + Send + Sync + 'static {
```
Comment (hlbarber) [inline rust-runtime/aws-smithy-http-server/src/routing/into_make_service_with_connect_info.rs:103]: Is `Clone` required here?

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/routing/into_make_service_with_connect_info.rs:103]: What happens if we remove `Send + Sync + 'static` too?

**Review by hlbarber — APPROVED** [review]

**Comment by hlbarber** [comment]
At some point perhaps it makes sense to re-home `IntoMakeService` and `IntoMakeServiceWithConnectInfo` into a different module. The reason being, these aren't _really_ to do with `routing` at all.

File: rust-runtime/aws-smithy-http-server/src/routing/into_make_service_with_connect_info.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+/// Extractor for getting connection information produced by a `Connected`.
+///
+/// Note this extractor requires you to use
+/// [`Router::into_make_service_with_connect_info`] to run your app
+/// otherwise it will fail at runtime.
+///
+/// See [`Router::into_make_service_with_connect_info`] for more details.
+///
+/// [`Router::into_make_service_with_connect_info`]: crate::routing::Router::into_make_service_with_connect_info
+#[derive(Clone, Copy, Debug)]
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/routing/into_make_service_with_connect_info.rs:150]: Why is this `Copy`?

**Comment by 82marbag** [comment]
> Can we add a code example binary under `examples` showing how to extract the peer's IP address? We've had this request in the past; it'd be nice to have an example to link to.

Added

File: rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/bin/pokemon-service-conninfo.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+struct Args {
+    /// Hyper server bind address.
+    #[clap(short, long, action, default_value = "127.0.0.1")]
+    address: String,
+    /// Hyper server bind port.
+    #[clap(short, long, action, default_value = "13734")]
+    port: u16,
+}
+
+/// Retrieves the users storage. No authentication required for locals.
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/bin/pokemon-service-conninfo.rs:22]:
```suggestion
/// Retrieves the user's storage. No authentication required for locals.
```

File: rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/bin/pokemon-service-conninfo.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+pub async fn get_storage_with_local_approved(
+    input: pokemon_service_server_sdk::input::GetStorageInput,
+    conn_info: aws_smithy_http_server::Extension<aws_smithy_http_server::routing::ConnectInfo<std::net::SocketAddr>>,
```
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/examples/pokemon-service/src/bin/pokemon-service-conninfo.rs:25]: nit: I'd just use `connect_info` here and in the filename.

**Review by david-perez — APPROVED** [review]

**Comment by 82marbag** [comment]
Blocked, pending https://github.com/awslabs/smithy-rs/pull/1980

**Pull request merged**